README.md
# GELF Appender for Apache Log4j 2
[![Build Status](https://travis-ci.org/Graylog2/log4j2-gelf.svg?branch=master)](https://travis-ci.org/Graylog2/log4j2-gelf)

This appender for [Apache Log4j 2](https://logging.apache.org/log4j/2.x/) logs messages to a GELF server like [Graylog2](http://www.graylog2.org) or [logstash](http://logstash.net).

It's using the official [GELF Java client](https://graylog2.github.io/gelfclient/) to connect to a remote server.

You can specify the following parameters for the GELF appender in the `log4j2.xml` configuration file:

* `name`
  * The reference name of the Appender
* `server` (default: `localhost`)
  * The host name or IP address of the GELF server
* `port` (default: `12201`)
  * The port the GELF server is listening on
* `hostName` (default: the local host name or `localhost` if it couldn't be detected)
  * The host name of the machine generating the logs
* `protocol` (default: `UDP`)
  * The transport protocol to use
* `includeSource` (default: `true`)
  * Whether the source of the log message should be included
* `includeThreadContext` (default: `true`)
  * Whether the contents of the [ThreadContext](https://logging.apache.org/log4j/2.x/manual/thread-context.html) should be included
* `includeStackTrace` (default: `true`)
  * Whether a full stack trace should be included
* `queueSize` (default: `512`)
  * The size of the internally used queue
* `connectTimeout` (default: `1000`)
  * The connection timeout for TCP connections in milliseconds
* `reconnectDelay` (default: `500`)
  * The time to wait between reconnects in milliseconds
* `sendBufferSize` (default: `-1`)
  * The size of the socket send buffer in bytes. A size of -1 deactivates the send buffer
* `tcpNoDelay` (default: `false`)
  * Whether Nagle's algorithm should be used for TCP connections
* `tcpKeepAlive` (default: `false`)
  * Whether to try keeping alive TCP connections.
* `filter`
  * A [Filter](https://logging.apache.org/log4j/2.x/manual/filters.html) to determine if the event should be handled by this Appender
* `layout` (default: `"%m%n"`)
  * The [Layout](https://logging.apache.org/log4j/2.x/manual/layouts.html) to use to format the LogEvent
* `ignoreExceptions`
  * The default is `true`, causing exceptions encountered while appending events to be internally logged and then ignored. When set to `false` exceptions will be propagated to the caller, instead. Must be set to `false` when wrapping this Appender in a `FailoverAppender`.
* `additionalFields`
  * Comma-delimited list of key=value pairs to be included in every message

## Log4j2.xml example

    <configuration status="OFF" packages="org.graylog2.log4j2">
        <appenders>
            <GELF name="gelfAppender" server="graylog2.example.com" port="12201" hostName="appserver01.example.com" additionalFields="foo=bar"/>
        </appenders>
        <loggers>
            <root level="info">
                <appender-ref ref="gelfAppender"/>
            </root>
        </loggers>
    </configuration>


## Java code example

    Logger logger = LogManager.getLogger("test");
    ThreadContext.put("userId", "testUser");
    logger.info("Hello World");


# Versions

| GELF Appender for Apache Log4j 2 | Release date |
| -------------------------------- | ------------ |
| 1.0.0                            | 29-Sep-2014  |
| 1.0.1                            | 21-Oct-2014  |
| 1.0.2                            | 09-Feb-2015  |
| 1.0.3                            | 11-Feb-2015  |
| 1.1.0                            | 16-Jul-2015  |

This appender uses GELF Java client 1.3.0.


# Installation

Maven coordinates

    <dependencies>
        <dependency>
            <groupId>org.graylog2.log4j2</groupId>
            <artifactId>log4j2-gelf</artifactId>
            <version>1.1.0</version>
        </dependency>
    </dependencies>


# License

GELF Appender for Apache Log4j 2

Copyright (C) 2014 TORCH GmbH; 2015 Graylog, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# log4j-gelf Changelog

## 1.1.0

* Add support for TCP keepalive
* Upgrade to gelfclient 1.3.0
* Provide a shaded JAR containing all dependencies


## 1.0.3

* Additional static fields were only included if message contained an exception


## 1.0.2

* Add setting for additional static fields to configuration
* Add missing package attribute in configuration example in README.md


## 1.0.1

* Actually start GelfAppender


# 1.0.0

* First public release


src/test/java/org/graylog2/log4j2/GelfAppenderTest.java

package org.graylog2.log4j2;


import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.Marker;
import org.apache.logging.log4j.MarkerManager;
import org.apache.logging.log4j.ThreadContext;
import org.junit.AfterClass;
import org.junit.Test;


public class GelfAppenderTest {
    @Test
    public void testLog() {
        final Logger logger = LogManager.getLogger("test");
        logger.info("Hello World");
    }


    @Test
    public void testMarker() {
        final Logger logger = LogManager.getLogger("test");
        final Marker parent = MarkerManager.getMarker("PARENT");
        final Marker marker = MarkerManager.getMarker("TEST").addParents(parent);
        logger.info(marker, "Hello World");
    }


    @Test
    public void testException() {
        final Logger logger = LogManager.getLogger("test");

        try {
            throw new Exception("Test");
        } catch (Exception e) {
            e.fillInStackTrace();
            logger.error("Hello World", e);
        }
    }


    @Test
    public void testThreadContext() {
        final Logger logger = LogManager.getLogger("test");

        ThreadContext.push("Message only");
        ThreadContext.push("int", 1);
        ThreadContext.push("int-long-string", 1, 2l, "3");
        ThreadContext.put("key", "value");

        logger.info("Hello World");

        ThreadContext.clearAll();

    }


    @AfterClass
    public static void shutdown() throws InterruptedException {
        //need to wait to hope the underlying gelf client pushes the messages.
        Thread.sleep(500);
    }
}


src/main/java/org/graylog2/log4j2/GelfAppender.java
package org.graylog2.log4j2;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.Marker;
import org.apache.logging.log4j.core.Filter;
import org.apache.logging.log4j.core.Layout;
import org.apache.logging.log4j.core.LogEvent;
import org.apache.logging.log4j.core.appender.AbstractAppender;
import org.apache.logging.log4j.core.appender.AppenderLoggingException;
import org.apache.logging.log4j.core.config.plugins.Plugin;
import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
import org.apache.logging.log4j.core.config.plugins.PluginElement;
import org.apache.logging.log4j.core.config.plugins.PluginFactory;
import org.apache.logging.log4j.core.layout.PatternLayout;
import org.apache.logging.log4j.core.net.Severity;
import org.apache.logging.log4j.status.StatusLogger;
import org.graylog2.gelfclient.GelfConfiguration;
import org.graylog2.gelfclient.GelfMessageBuilder;
import org.graylog2.gelfclient.GelfMessageLevel;
import org.graylog2.gelfclient.GelfTransports;
import org.graylog2.gelfclient.transport.GelfTransport;

import java.io.Serializable;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.UnknownHostException;
import java.util.Collections;
import java.util.Formatter;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Plugin(name = "GELF", category = "Core", elementType = "appender", printObject = true)
public class GelfAppender extends AbstractAppender {
    private static final long serialVersionUID = 4796033328540158817L;

    private static final Logger LOG = StatusLogger.getLogger();

    private final GelfConfiguration gelfConfiguration;
    private final String hostName;
    private final boolean includeSource;
    private final boolean includeThreadContext;
    private final boolean includeStackTrace;
    private final Map<String, Object> additionalFields;

    private GelfTransport client;

    protected GelfAppender(final String name,
                           final Layout<? extends Serializable> layout,
                           final Filter filter,
                           final boolean ignoreExceptions,
                           final GelfConfiguration gelfConfiguration,
                           final String hostName,
                           final boolean includeSource,
                           final boolean includeThreadContext,
                           final boolean includeStackTrace,
                           String additionalFields) {
        super(name, filter, layout, ignoreExceptions);
        this.gelfConfiguration = gelfConfiguration;
        this.hostName = hostName;
        this.includeSource = includeSource;
        this.includeThreadContext = includeThreadContext;
        this.includeStackTrace = includeStackTrace;

        if (null != additionalFields && !additionalFields.isEmpty()) {
            this.additionalFields = new HashMap<>();

            try {
                String[] values = additionalFields.split(",");
                for (String s : values) {
                    String[] nvp = s.split("=");
                    this.additionalFields.put(nvp[0], nvp[1]);
                }
            } catch (Exception e) {
                LOGGER.warn("Failed to read additional fields.", e);
            }
        } else {
            this.additionalFields = Collections.emptyMap();
        }
    }

    @Override
    public void append(LogEvent event) {
        final GelfMessageBuilder builder = new GelfMessageBuilder(event.getMessage().getFormattedMessage(), hostName)
                .timestamp(event.getTimeMillis() / 1000d)
                .level(GelfMessageLevel.fromNumericLevel(Severity.getSeverity(event.getLevel()).getCode()))
                .additionalField("loggerName", event.getLoggerName())
                .additionalField("threadName", event.getThreadName());

        final Marker marker = event.getMarker();
        if (marker != null) {
            builder.additionalField("marker", marker.getName());
        }

        if (includeThreadContext) {
            for (Map.Entry<String, String> entry : event.getContextMap().entrySet()) {
                builder.additionalField(entry.getKey(), entry.getValue());
            }

            final List<String> contextStack = event.getContextStack().asList();
            if (contextStack != null && !contextStack.isEmpty()) {
                builder.additionalField("contextStack", contextStack.toString());
            }
        }

        final StackTraceElement source = event.getSource();
        if (includeSource && source != null) {
            builder.additionalField("sourceFileName", source.getFileName());
            builder.additionalField("sourceMethodName", source.getMethodName());
            builder.additionalField("sourceClassName", source.getClassName());
            builder.additionalField("sourceLineNumber", source.getLineNumber());
        }

        @SuppressWarnings("all")
        final Throwable thrown = event.getThrown();
        if (includeStackTrace && thrown != null) {
            final StringBuilder stackTraceBuilder = new StringBuilder();
            for (StackTraceElement stackTraceElement : thrown.getStackTrace()) {
                new Formatter(stackTraceBuilder).format("%s.%s(%s:%d)%n",
                        stackTraceElement.getClassName(),
                        stackTraceElement.getMethodName(),
                        stackTraceElement.getFileName(),
                        stackTraceElement.getLineNumber());
            }

            builder.additionalField("exceptionClass", thrown.getClass().getCanonicalName());
            builder.additionalField("exceptionMessage", thrown.getMessage());
            builder.additionalField("exceptionStackTrace", stackTraceBuilder.toString());

            builder.fullMessage(event.getMessage().getFormattedMessage() + "\n\n" + stackTraceBuilder.toString());
        }
        
        if (!additionalFields.isEmpty()) {
        	builder.additionalFields(additionalFields);
        }

        try {
            client.send(builder.build());
        } catch (Exception e) {
            throw new AppenderLoggingException("failed to write log event to GELF server: " + e.getMessage(), e);
        }
    }

    @Override
    public void start() {
        super.start();
        client = GelfTransports.create(gelfConfiguration);
    }

    @Override
    public void stop() {
        super.stop();
        client.stop();
    }

    @Override
    public String toString() {
        return GelfAppender.class.getSimpleName() + "{"
                + "name=" + getName()
                + ",server=" + gelfConfiguration.getRemoteAddress().getHostName()
                + ",port=" + gelfConfiguration.getRemoteAddress().getPort()
                + ",protocol=" + gelfConfiguration.getTransport().toString()
                + ",hostName=" + hostName
                + ",queueSize=" + gelfConfiguration.getQueueSize()
                + ",connectTimeout=" + gelfConfiguration.getConnectTimeout()
                + ",reconnectDelay=" + gelfConfiguration.getReconnectDelay()
                + ",sendBufferSize=" + gelfConfiguration.getSendBufferSize()
                + ",tcpNoDelay=" + gelfConfiguration.isTcpNoDelay()
                + ",tcpKeepAlive=" + gelfConfiguration.isTcpKeepAlive()
                + "}";
    }

    /**
     * Factory method for creating a {@link GelfTransport} provider within the plugin manager.
     *
     * @param name                 The name of the Appender.
     * @param filter               A Filter to determine if the event should be handled by this Appender.
     * @param layout               The Layout to use to format the LogEvent defaults to {@code "%m%n"}.
     * @param ignoreExceptions     The default is {@code true}, causing exceptions encountered while appending events
     *                             to be internally logged and then ignored. When set to {@code false} exceptions will
     *                             be propagated to the caller, instead. Must be set to {@code false} when wrapping this
     *                             Appender in a {@link org.apache.logging.log4j.core.appender.FailoverAppender}.
     * @param server               The server name of the GELF server, defaults to {@code localhost}.
     * @param port                 The port the GELF server is listening on, defaults to {@code 12201}.
     * @param hostName             The host name of the machine generating the logs, defaults to local host name
     *                             or {@code localhost} if it couldn't be detected.
     * @param protocol             The transport protocol to use, defaults to {@code UDP}.
     * @param queueSize            The size of the internally used queue, defaults to {@code 512}.
     * @param connectTimeout       The connection timeout for TCP connections in milliseconds, defaults to {@code 1000}.
     * @param reconnectDelay       The time to wait between reconnects in milliseconds, defaults to {@code 500}.
     * @param sendBufferSize       The size of the socket send buffer in bytes, defaults to {@code -1} (deactivate).
     * @param tcpNoDelay           Whether Nagle's algorithm should be used for TCP connections, defaults to {@code false}.
     * @param tcpKeepAlive         Whether to try keeping alive TCP connections, defaults to {@code false}.
     * @param includeSource        Whether the source of the log message should be included, defaults to {@code true}.
     * @param includeThreadContext Whether the contents of the {@link org.apache.logging.log4j.ThreadContext} should be included, defaults to {@code true}.
     * @param includeStackTrace    Whether a full stack trace should be included, defaults to {@code true}.
     * @param additionalFields     Additional static comma-delimited key=value pairs that will be added to every log message.
     * @return a new GELF provider
     */
    @PluginFactory
    public static GelfAppender createGelfAppender(@PluginElement("Filter") Filter filter,
                                                  @PluginElement("Layout") Layout<? extends Serializable> layout,
                                                  @PluginAttribute(value = "name") String name,
                                                  @PluginAttribute(value = "ignoreExceptions", defaultBoolean = true) Boolean ignoreExceptions,
                                                  @PluginAttribute(value = "server", defaultString = "localhost") String server,
                                                  @PluginAttribute(value = "port", defaultInt = 12201) Integer port,
                                                  @PluginAttribute(value = "protocol", defaultString = "UDP") String protocol,
                                                  @PluginAttribute(value = "hostName") String hostName,
                                                  @PluginAttribute(value = "queueSize", defaultInt = 512) Integer queueSize,
                                                  @PluginAttribute(value = "connectTimeout", defaultInt = 1000) Integer connectTimeout,
                                                  @PluginAttribute(value = "reconnectDelay", defaultInt = 500) Integer reconnectDelay,
                                                  @PluginAttribute(value = "sendBufferSize", defaultInt = -1) Integer sendBufferSize,
                                                  @PluginAttribute(value = "tcpNoDelay", defaultBoolean = false) Boolean tcpNoDelay,
                                                  @PluginAttribute(value = "tcpKeepAlive", defaultBoolean = false) Boolean tcpKeepAlive,
                                                  @PluginAttribute(value = "includeSource", defaultBoolean = true) Boolean includeSource,
                                                  @PluginAttribute(value = "includeThreadContext", defaultBoolean = true) Boolean includeThreadContext,
                                                  @PluginAttribute(value = "includeStackTrace", defaultBoolean = true) Boolean includeStackTrace,
                                                  @PluginAttribute(value = "additionalFields") String additionalFields) {
        if (name == null) {
            LOGGER.error("No name provided for ConsoleAppender");
            return null;
        }
        if (layout == null) {
            layout = PatternLayout.createDefaultLayout();
        }
        if (!"UDP".equalsIgnoreCase(protocol) && !"TCP".equalsIgnoreCase(protocol)) {
            LOG.warn("Invalid protocol {}, falling back to UDP", protocol);
            protocol = "UDP";
        }
        if (hostName == null || hostName.trim().isEmpty()) {
            try {
                hostName = InetAddress.getLocalHost().getHostName();
            } catch (UnknownHostException e) {
                LOG.warn("Couldn't detect local host name, falling back to \"localhost\"");
                hostName = "localhost";
            }
        }

        final InetSocketAddress serverAddress = new InetSocketAddress(server, port);
        final GelfTransports gelfProtocol = GelfTransports.valueOf(protocol.toUpperCase());
        final GelfConfiguration gelfConfiguration = new GelfConfiguration(serverAddress)
                .transport(gelfProtocol)
                .queueSize(queueSize)
                .connectTimeout(connectTimeout)
                .reconnectDelay(reconnectDelay)
                .sendBufferSize(sendBufferSize)
                .tcpNoDelay(tcpNoDelay)
                .tcpKeepAlive(tcpKeepAlive);

        return new GelfAppender(name, layout, filter, ignoreExceptions, gelfConfiguration, hostName, includeSource,
                includeThreadContext, includeStackTrace, additionalFields);
    }
}
